### PR TITLE
Writing benchmark data at the end of the game

### PR DESCRIPTION
**Pull request by teinarss**
Updated the benchmark mode to handle all perfsamples and writing to disk at the end.

### CODE REVIEW

File: OpenRA.Game/Support/Benchmark.cs
Comment (abcdefg30) [inline OpenRA.Game/Support/Benchmark.cs]: Missing header.

File: OpenRA.Game/Support/Benchmark.cs
Comment (abcdefg30) [inline OpenRA.Game/Support/Benchmark.cs]: Please drop braces around single-lines.

**Comment by teinarss** [comment]
Update: Fixed the issues

**Comment by teinarss** [comment]
On hold! Got some new ideas that i want to implement in this PR before merge.

**Comment by pchote** [comment]
Do you plan to revisit this soon? Might be best to close this if not.

**Comment by teinarss** [comment]
### Updated
Added support to launch a map directly and exit when the game is over

The use case is to support automate running bechmarks several times for getting better data.

File: OpenRA.Mods.Common/LoadScreens/BlankLoadScreen.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/LoadScreens/BlankLoadScreen.cs]: This doesn't look correct.

**Comment by teinarss** [comment]
Updated

**Comment by abcdefg30** [comment]
Needs a rebase.

**Comment by teinarss** [comment]
Rebased.
Did some more changes after developing a more automated system to run it several times.

File: OpenRA.Game/Support/LaunchArguments.cs
Comment (pchote) [inline OpenRA.Game/Support/LaunchArguments.cs]: Can we please merge this and `ExitOnEnd` into a single `BenchmarkMap` argument?

File: OpenRA.Game/Game.cs
Comment (pchote) [inline OpenRA.Game/Game.cs]: IMO the interface here isn't too useful - we can use `Benchmark` directly and default to `null` for normal games.

File: OpenRA.Game/Game.cs
Comment (pchote) [inline OpenRA.Game/Game.cs]: This can then simplify to
```
if (benchmark != null)
{
    benchmark.Write();
    Exit();
}
```

**Review by pchote — COMMENTED** [review]
Some thoughts on streamlining this:

**Comment by teinarss** [comment]
I would like to have the separate the map with benchmark arguments so that
we can use it with replays, so let's combine benchmark with exitonend and
then keep the map separated.

**Comment by pchote** [comment]
How important is the ability to set the benchmark output folder?
Another option could be to have just `Benchmark` which takes either a map uid or replay.

**Comment by teinarss** [comment]
Its most for convenience, im using it to mark the data dumps like this:
baseline-0-Pathfinder.csv

**Comment by teinarss** [comment]
Updated and rebased

File: OpenRA.Game/Game.cs
Comment (pchote) [inline OpenRA.Game/Game.cs]: `map.Uid` NREs if the map isn't found.

File: OpenRA.Game/Game.cs
Comment (pchote) [inline OpenRA.Game/Game.cs]: We generally specify maps by their UID or path.
Can we please change this to
```
var path = Platform.ResolvePath(launchMap);
var map = ModData.MapCache.SingleOrDefault(m => m.Uid == launchMap) ??
	ModData.MapCache.SingleOrDefault(m => m.Package.Name == path);
``` 
?

File: OpenRA.Game/Support/LaunchArguments.cs
Comment (pchote) [inline OpenRA.Game/Support/LaunchArguments.cs]: This will also be really useful for people working on scripted maps, so good call not merging it with `Benchmark`.

Maybe this will also perk @MustaphaTR or @abcdefg30's interest to review this PR.

**Comment by teinarss** [comment]
updated!

File: OpenRA.Game/Support/Benchmark.cs
Comment (pchote) [inline OpenRA.Game/Support/Benchmark.cs]: We already have an extension method for this!

File: OpenRA.Game/Support/Benchmark.cs
Comment (pchote) [inline OpenRA.Game/Support/Benchmark.cs]: This can simplify to
```csharp
foreach (var item in PerfHistory.Items)
	samples.GetOrAdd(item.Key).Add(new BenchmarkPoint(localTick, item.Value.LastValue));
```

**Review by pchote — COMMENTED** [review]
One last simplification then LGTM

**Comment by teinarss** [comment]
Updated

**Review by pchote — DISMISSED** [review]

File: OpenRA.Game/Support/LaunchArguments.cs
Comment (abcdefg30) [reply OpenRA.Game/Support/LaunchArguments.cs]: Fwiw, editing a Lua script does not change the hash (or path), but one can just use the restart map function to reload the script anyway.

**Review by abcdefg30 — COMMENTED** [review]
Looks good to me. From the IRC discussion: We want to write benchmark data when using "Abort Mission", too.

**Comment by teinarss** [comment]
Updated: Added support to abort mission, using World.Dispose() required a manuel input when using a scripted map with a timed win condition.

**Review by pchote — APPROVED** [review]

**Review by abcdefg30 — APPROVED** [review]

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/e3785d7cfe337fbce7d595749ad9a8a7c39a5cc6...547dba575bff35b612adad46f7cceea6bf5f4947)

**Pull request merged**
